ui/pages/setup_secure.tsx

```diff
@@ -44,29 +40,30 @@ const SetupSecure: NextPage = () => {
     // Otherwise we initialize it
     kratos
       .createBrowserSettingsFlow({
-        returnTo: returnTo ? String(returnTo) : undefined,
+        returnTo: return_to
+          ? return_to.toString()
+          : window.location.pathname.replace(`setup_secure`, `setup_complete`),
       })
-      .then(({ data }) => {
-        if (data.request_url !== undefined) {
-          const pwParam = pwChanged
-            ? `&pw_changed=${pwChanged.toString()}`
-            : "";
-          window.location.href = `./setup_secure?flow=${data.id}${pwParam}`;
-          return;
-        }
+      .then(async ({ data }) => {
+        const pwParam = pwChanged ? pwChanged.toString() : "false";
+        router.query.flow = data.id;
+        router.query.pw_changed = pwParam;
+
+        await router.replace(
+          {
+            pathname: window.location.pathname,
+            query: router.query,
+          },
+          undefined,
+          { shallow: true },
+        );
```

Comment:
Did you test the flow with various combinations? Like 1. coming from the password change page or 2. hitting this page directly with an authenticated or 3. unauthenticated session? I think there was a reason we didn't use `router.replace` before, but I might be wrong.

Reply:
Thanks for checking. In my opinion, (3) should redirect to log in. Which it does on staging, but then you get to the error page, because we don't allow login when arriving without a valid login challenge or flow ID. Not sure if we can fix it here, or should just report it as a bug?

Reply:
The (3) case was likely handled by the removed lines 61-66 below, maybe they should be restored?

Reply:
I don't think that ever worked (because of what you said on your comment about the login challenge). I would like to remove it for now and we can fix it properly in another PR.